### PR TITLE
Don't calculate the Finished MAC twice

### PR DESCRIPTION
**Pull request by mattcaswell**
In <= TLSv1.2 a Finished message always comes immediately after a CCS
except in the case of NPN where there is an additional message between
the CCS and Finished. Historically we always calculated the Finished MAC
when we processed the CCS. However to deal with NPN we also calculated it
when we receive the Finished message. Really this should only have been
done if we hand negotiated NPN.

This simplifies the code to only calculate the MAC when we receive the
Finished. In 1.1.1 we need to do it this way anyway because there is no
CCS (except in middlebox compat mode) in TLSv1.3.

Coincidentally, this commit also fixes the fact that no-nextprotoneg does
not currently work in master.

### CODE REVIEW

File: ssl/statem/statem_lib.c [hunk 1]
```diff
@@ -646,12 +646,11 @@ MSG_PROCESS_RETURN tls_process_key_update(SSL *s, PACKET *pkt)
     return MSG_PROCESS_FINISHED_READING;
 }
 
-#ifndef OPENSSL_NO_NEXTPROTONEG
 /*
  * ssl3_take_mac calculates the Finished MAC for the handshakes messages seen
  * to far.
  */
-static void ssl3_take_mac(SSL *s)
+void ssl3_take_mac(SSL *s)
```
Comment (dot-asm) [inline ssl/statem/statem_lib.c]: Removed code was checking return value from final_finish_mac, this one doesn't. Is it checked elsewhere?

Comment (mattcaswell) [reply ssl/statem/statem_lib.c]: Good point. I added the missing check.

**Comment by mattcaswell** [comment]
New commit pushed as fixup (to be squashed). Similarly for #5286.

File: ssl/statem/statem_lib.c [hunk 2]
```diff
@@ -1227,14 +1226,15 @@ int tls_get_message_body(SSL *s, size_t *len)
...
 
-#ifndef OPENSSL_NO_NEXTPROTONEG
     /*
      * If receiving Finished, record MAC of prior handshake messages for
      * Finished verification.
      */
-    if (*s->init_buf->data == SSL3_MT_FINISHED)
-        ssl3_take_mac(s);
-#endif
+    if (*s->init_buf->data == SSL3_MT_FINISHED && !ssl3_take_mac(s)) {
```
Comment (dot-asm) [inline ssl/statem/statem_lib.c]: I found `*s->init_buf->data` a bit hard to read. `*(s->init_buf->data)` or `s->init_buf->data[0]` would be better. This is somewhat of a side note, so "bug off" is acceptable answer, in which case I'd be ready to approve...

Comment (mattcaswell) [reply ssl/statem/statem_lib.c]: I went with `*(s->init_buf->data)`. Both here and in #5286.

**Comment by mattcaswell** [comment]
Updated commit pushed.

**Comment by mattcaswell** [comment]
Pushed. Thanks.

**Pull request closed without merging**